File: src/main/java/com/example/SecurityConfig.java
package com.example;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.annotation.Order;
import org.springframework.security.authentication.TestingAuthenticationToken;
import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity;
import org.springframework.security.config.web.server.SecurityWebFiltersOrder;
import org.springframework.security.config.web.server.ServerHttpSecurity;
import org.springframework.security.core.context.ReactiveSecurityContextHolder;
import org.springframework.security.web.server.SecurityWebFilterChain;
import org.springframework.security.web.server.context.SecurityContextServerWebExchangeWebFilter;
import org.springframework.security.web.server.util.matcher.ServerWebExchangeMatcher;
import org.springframework.web.server.WebFilter;


/**
 * @author Rob Winch
 */
@Configuration
public class SecurityConfig {

	@Bean
	SecurityContextServerWebExchangeWebFilter reactorContextWebFilter() {
		return new SecurityContextServerWebExchangeWebFilter();
	}

	@Bean
	@Order(0)
	WebFilter simpleFilter() {
		return (e,c) ->  c
			.filter(e)
			.subscriberContext(ReactiveSecurityContextHolder.withAuthentication(new TestingAuthenticationToken("the-username", "p", "USER")));
	}
}

File: src/main/kotlin/com/example/SecurityConfig.kt
package com.example

import org.springframework.context.annotation.Bean
import org.springframework.security.authentication.ReactiveAuthenticationManager
import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity
import org.springframework.security.config.web.server.SecurityWebFiltersOrder
import org.springframework.security.config.web.server.ServerHttpSecurity
import org.springframework.security.core.Authentication
import org.springframework.security.core.userdetails.User
import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken
import org.springframework.security.web.server.SecurityWebFilterChain
import org.springframework.security.web.server.authentication.AuthenticationWebFilter
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Mono
import reactor.core.publisher.toMono

@EnableWebFluxSecurity
class SecurityConfig {

    @Bean
    fun springSecurityFilterChain(http: ServerHttpSecurity): SecurityWebFilterChain {
        return http
            .csrf()
            .disable()
            .addFilterAt(simpleFilter(), SecurityWebFiltersOrder.AUTHENTICATION)
            .authorizeExchange()
            .anyExchange().authenticated()
            .and()
            .build()
    }

    fun simpleFilter() = AuthenticationWebFilter(NoopAuthenticationManager).apply {
        setAuthenticationConverter(SimpleAuthenticationConverter)
    }

    object SimpleAuthenticationConverter : (ServerWebExchange) -> Mono<Authentication> {
        override fun invoke(exchange: ServerWebExchange): Mono<Authentication> {
            val user = User("test", "N/A", emptyList())
            return PreAuthenticatedAuthenticationToken(user, user.password, user.authorities).toMono()
        }
    }

    object NoopAuthenticationManager : ReactiveAuthenticationManager {

        override fun authenticate(authentication: Authentication): Mono<Authentication> {
            return if (authentication is PreAuthenticatedAuthenticationToken)
                Mono.just(authentication)
            else
                Mono.error(IllegalStateException("Only X.509 Auth accepted."))
        }

    }

}
